pkg/informer/informer.go

```diff
@@ -126,108 +158,150 @@ func NewDynamicDiscoverySharedInformerFactory(
 
 	f.handlers.Store([]GVREventHandler{})
 
-	// Add an index function that indexes a CRD by its group/version/resource.
-	if err := crdInformer.Informer().AddIndexers(cache.Indexers{
-		byGroupVersionResourceIndex: byGroupVersionResourceIndexFunc,
-	}); err != nil {
-		return nil, err
-	}
-
-	// Any time a CRD event comes in, let StartWorker() know about it
-	notifyUpdateNeeded := func() {
-		select {
-		case f.updateCh <- struct{}{}:
-			klog.V(4).InfoS("Enqueued update notification for dynamic informer recalculation")
-		default:
-			klog.V(5).InfoS("Dropping update notification for dynamic informer recalculation because a notification is already pending")
-		}
-	}
-
-	crdIsEstablished := func(obj interface{}) bool {
-		crd, ok := obj.(*apiextensionsv1.CustomResourceDefinition)
-		if !ok {
-			utilruntime.HandleError(fmt.Errorf("unable to determine if CRD is established - got type %T instead", obj))
-			return false
-		}
-
-		return apihelpers.IsCRDConditionTrue(crd, apiextensionsv1.Established)
-	}
-
-	// When CRDs change, send a notification that we might need to add/remove informers.
-	crdInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
-		AddFunc: func(obj interface{}) {
-			if crdIsEstablished(obj) {
-				notifyUpdateNeeded()
-			}
-		},
-		UpdateFunc: func(oldObj, newObj interface{}) {
-			oldEstablished := crdIsEstablished(oldObj)
-			newEstablished := crdIsEstablished(newObj)
-			if newEstablished || oldEstablished != newEstablished {
-				notifyUpdateNeeded()
+	changes := gvrSource.Changes()
+	go func() {
+		for change := range changes {
+			select {
+			case f.updateCh <- change:
+				klog.V(4).InfoS("Enqueued update notification for dynamic informer recalculation")
+			default:
+				klog.V(5).InfoS("Dropping update notification for dynamic informer recalculation because a notification is already pending")
 			}
-		},
-		DeleteFunc: func(obj interface{}) {
-			notifyUpdateNeeded()
-		},
-	})
+		}
+	}()
 
 	return f, nil
 }
 
-func byGroupVersionResourceKeyFunc(group, version, resource string) string {
-	return fmt.Sprintf("%s/%s/%s", group, version, resource)
+// NewScopedDiscoveringDynamicSharedInformerFactory returns a factory for
+// cluster-unaware (standard Kube) shared informers that discovers new types
+// and informs on updates to resources of those types.
+// It receives the GVR-related information by delegating to a GVRSource.
+func NewScopedDiscoveringDynamicSharedInformerFactory(
```

Comment:
We don't need this constructor - what keeps callers from just using the version with generics and the appropriate GVRSource?

Reply:
This constructor includes the implementation of the `newInformer()` function. It is obviously specific to the cluster-unaware case, but still part of the ddsif logic. So I don't see how it would make sense to have every caller define it.
IMO it makes sense to be encapsulated in this package.

Reply:
Can we find a factoring that doesn't require these extra constructors?

Reply:
I tried, but I might have missed something.

Reply:
OK, let's keep it this way then!